### PR TITLE
DEV: Add integration specs for Github login

### PR DESCRIPTION
**Pull request by OsamaSayegh**
Internal topic: t/82084.

### CODE REVIEW

File: spec/integration/github_omniauth_spec.rb [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(cookies[:_t]).to be_blank
+  end
+
+  it "favors the primary email over secondary emails when they're all verified" do
+    post "/auth/github"
```
Comment (tgxworld) [inline spec/integration/github_omniauth_spec.rb:203]: Is there a response we're expecting here?

File: spec/integration/github_omniauth_spec.rb [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(
+      UserAuthToken.lookup(request.cookie_jar.encrypted[:_t]["token"]).user.id
+    ).to eq(user2.id)
+  end
+
+  it "doesn't match a non-primary email if it's not verified" do
+    post "/auth/github"
```
Comment (tgxworld) [inline spec/integration/github_omniauth_spec.rb:174]: Same as https://github.com/discourse/discourse/pull/18808/files#r1009938805

File: spec/integration/github_omniauth_spec.rb [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(cookies[:_t]).to be_blank
+  end
+
+  it "matches a non-primary email if it's verified and the primary email isn't" do
+    post "/auth/github"
```
Comment (tgxworld) [inline spec/integration/github_omniauth_spec.rb:143]: same as https://github.com/discourse/discourse/pull/18808/files#r1009938805

File: spec/integration/github_omniauth_spec.rb [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+          }
+        ),
+        headers: {
+          "Content-Type" => "application/json"
+        }
+      )
+  end
+
+  it "doesn't sign in anyone if none of the emails from github are verified" do
+    post "/auth/github"
```
Comment (tgxworld) [inline spec/integration/github_omniauth_spec.rb:114]: Same as https://github.com/discourse/discourse/pull/18808/files#r1009938805

File: spec/integration/github_omniauth_spec.rb [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      ]
+    )
+
+    post "/auth/github/callback", params: {
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(cookies[:_t]).to be_blank
```
Comment (tgxworld) [inline spec/integration/github_omniauth_spec.rb:139]: Small note you can rely on `session[:current_user_id]` to determine if a user is logged in or not

https://github.com/discourse/discourse/blob/d1c5529aadda4942e4981f4fd9f7f9699dc9e231/spec/rails_helper.rb#L405

File: spec/integration/github_omniauth_spec.rb [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+    post "/auth/github/callback", params: {
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(
+      UserAuthToken.lookup(request.cookie_jar.encrypted[:_t]["token"]).user.id
+    ).to eq(user2.id)
```
Comment (tgxworld) [inline spec/integration/github_omniauth_spec.rb:170]: I'm not quite familiar with the code here but what does creating the `UserAuthToken` do?

File: spec/integration/github_omniauth_spec.rb (hunk 1 above)
Comment (OsamaSayegh) [reply spec/integration/github_omniauth_spec.rb:207]: This request redirects to github and sets up some state in the session that are needed for the later request. I've added a couple of assertions for ensuring this requests 302 redirects to github.

File: spec/integration/github_omniauth_spec.rb (hunk 5 above)
Comment (OsamaSayegh) [reply spec/integration/github_omniauth_spec.rb]: Awesome, thanks! Updated the specs to use `session[:current_user_id]`.

File: spec/integration/github_omniauth_spec.rb (hunk 6 above)
Comment (OsamaSayegh) [reply spec/integration/github_omniauth_spec.rb]: When a user is logged in, a `UserAuthToken` record is created which contains the user's auth cookie (hashed of course) and some information about the user's session. I'm asserting here that the auth cookie that is added by the request belongs to the correct user. However, I've now switched to use `session[:current_user_id]` for my assertion which is simpler/shorter.

**Review by tgxworld — APPROVED** [review]

**Pull request merged**